### PR TITLE
Fix clipping

### LINKED ISSUE #325: Group clipping into layer causes RecursionError
**Describe the bug**
Compositing results in a stack overflow when attempting to render set of layers containing a clipping group.

This stems from the fact that when opening, the group ("clipping" in the example file) is actually clipping into the layer that's inside itself ("red" in the example filed).

**To Reproduce**
```python
from psd_tools import PSDImage
psd = PSDImage.open('test.psd')
psd[2] # Group('clipping' size=225x424)
psd[2][0].clip_layers # [PixelLayer('blue' size=225x424), Group('clipping' size=225x424)]
psd.composite(ignore_preview=True)
```

**Expected behavior**
The document should render without errors if there were no errors while opening.
Ideally, the file would also render as expected.

**File and screenshots**
This PSD file was created in Clip Studio Paint, which supports clipping groups into layers.
[example.zip](https://github.com/psd-tools/psd-tools/files/8957098/example.zip)

The group and layer order (the vertical red bar means the group/layer is clipping into the layer below):
layers

How Clip Studio Paint renders the file:
test

**Environment**
 - Python [3.10.4]
 - Version [v1.9.21]

**Additional notes**
I made some changes to the layer loading function to support this functionality, but I'm unsure if it would cause problems with other features or not. Here's the patch:
[recursive-clipping-layers.patch](https://github.com/psd-tools/psd-tools/files/8957688/recursive-clipping-layers.txt)

Compositing appears to produce the correct output as long as the document tree is properly parsed.

### PR DESCRIPTION
**Pull request by cromachina**
Fixes #325

### CODE REVIEW

**Comment by cromachina** [comment]
I have made it so that clipping on to a pass through layer causes no clipping to occur, which is the same behavior as SAIv2 and CSP. However it seems like this is not correct according to Photoshop, because it's causing a test to fail where a layer is clipped onto a pass through group.

**Comment by cromachina** [comment]
I came up with a better way to deal with clipping layers than storing a clipping layer list in each layer when working on my own compositor, by simply storing the previous non-clipping layer's alpha, however, I'm not sure if I can easily translate it to `psd-tools`  compositor:

https://github.com/cromachina/psd-export/blob/9dd91aa39384bd201b2b4b7ada98dd4d0248be35/composite.py#L78-L106

**Comment by kyamagu** [comment]
@cromachina What would be the test to check the correctness?

**Comment by cromachina** [comment]
@kyamagu I don't have Photoshop so it's hard for me to confirm, but the test `test_layers.test_clip_adjustment` seems to be the offending one. From what I can infer from this test, it seems like in PS you can clip on to a `PASS_THROUGH` layer, but in SAI, the clipping behavior is ignored. The former would look like "source atop" and the latter would look like "source over". Does this sound correct?

File: src/psd_tools/api/psd_image.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def __init__(self, data):
         self._record = data
         self._layers = []
         self._tagged_blocks = None
+        self._compatibility_mode = CompatibilityMode.DEFAULT
```
Comment (kyamagu) [inline src/psd_tools/api/psd_image.py:42]: Actually, there is `VersionInfo` field in the resource section of a PSD file, that states the file creator like Adobe Photoshop. Perhaps this part should attempt at guessing the default compatibility mode.

```
from psd_tools.constants import Resource
version_info = psd.image_resources.get_data(Resource.VERSION_INFO)
```

File: src/psd_tools/constants.py [hunk 2]
```diff
@@ -1,8 +1,20 @@
 """
 Various constants for psd_tools
 """
-from enum import Enum, IntEnum
+from enum import Enum, IntEnum, auto
 
+class CompatibilityMode(Enum):
```
Comment (kyamagu) [inline src/psd_tools/constants.py:6]: Presumably there should be fixture files to test those. Any possibility?

File: src/psd_tools/api/psd_image.py [hunk 3]
```diff
@@ -525,6 +540,38 @@ def _get_pattern(self, pattern_id):
                         return pattern
         return None
 
+    def _clear_clipping_layers(self):
+        for layer in self.descendants():
+            layer._clip_layers = []
+            layer._has_clip_target = True
+
+    def _compute_clipping_layers(self):
```
Comment (kyamagu) [inline src/psd_tools/api/psd_image.py:548]: Appreciate if there is a short comment describing the behavior difference between tools

File: src/psd_tools/api/psd_image.py (hunk 1 above)
Comment (cromachina) [reply src/psd_tools/api/psd_image.py:42]: I can try something like that in bit.

**Pull request merged**